src/java/picard/sam/revertsam/RevertSamUtil.java

```diff
@@ -0,0 +1,188 @@
+ */
+public class RevertSamUtil {
+
+    private final static Log log = Log.getInstance(RevertSamUtil.class);
+
+    public static List<ReadGroup> makeReadGroupList(
+            final SAMFileHeader inHeader,
+            final OutputMap outputMap,
+            final SAMFileHeader.SortOrder sortOrder,
+            final boolean removeAlignmentInformation) {
+
+        final List<ReadGroup> readGroupList = new ArrayList<ReadGroup>();
+        for (final SAMReadGroupRecord samReadGroup : inHeader.getReadGroups()) {
+            final SAMFileHeader samHeader = createOutHeaderForReadGroup(samReadGroup, inHeader, sortOrder, removeAlignmentInformation);
+            final String id = samReadGroup.getId();
+            final ReadGroup readGroup = new ReadGroup(id, samHeader, outputMap.get(id));
+            readGroupList.add(readGroup);
+        }
+        return readGroupList;
+    }
+
+    private static SAMFileHeader createOutHeaderForReadGroup(
+            final SAMReadGroupRecord samReadGroup,
+            final SAMFileHeader inHeader,
+            final SAMFileHeader.SortOrder sortOrder,
+            final boolean removeAlignmentInformation) {
+
+        final SAMFileHeader outHeader = new SAMFileHeader();
+        outHeader.addReadGroup(samReadGroup);
+        outHeader.setSortOrder(sortOrder);
+        if (!removeAlignmentInformation) {
+            outHeader.setSequenceDictionary(inHeader.getSequenceDictionary());
+            outHeader.setProgramRecords(inHeader.getProgramRecords());
+        }
+        return outHeader;
+    }
+
+    public static RevertSamWriter makeSingleOrPerReadGroupWriter(
+            final SAMFileHeader inHeader,
+            final SAMFileHeader outHeader,
+            final boolean presorted,
+            final boolean outputByReadGroup,
+            final List<ReadGroup> readGroups,
+            final File singleOutput) {
+
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+        if (outputByReadGroup) {
+            final Map<String, SAMFileWriter> writers = createWriterMap(readGroups, presorted, factory);
+            return new PerReadGroupWriter(writers);
+        }
+        final SAMFileWriter innerWriter = factory.makeSAMOrBAMWriter(outHeader, presorted, singleOutput);
+        return new SingleSamFileWriter(innerWriter);
+    }
+
+    private static Map<String, SAMFileWriter> createWriterMap(
+            final List<ReadGroup> readGroups,
+            final boolean presorted,
+            final SAMFileWriterFactory factory) {
+
+        final Map<String, SAMFileWriter> writerMap = new HashMap<String, SAMFileWriter>();
+        for (final ReadGroup readGroup : readGroups) {
+            final SAMFileWriter writer = factory.makeSAMOrBAMWriter(readGroup.getSamHeader(), presorted, readGroup.getOutput());
+            writerMap.put(readGroup.getId(), writer);
+        }
+        return writerMap;
+    }
+
+    public static SAMFileHeader createSingleOutHeader(
```

Comment:
It was confusing that this is called `createSingleOutHeader` yet its output is eventually used only by `makeSingleOrPerReadGroupWriter` and only when creating `innerWriter`.  Seems hard to follow.  Why not just put it in `createWriterMap`?